common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -514,12 +512,12 @@ public MergeStatuses finalizeShuffleMerge(FinalizeShuffleMerge msg) {
     AppShuffleInfo appShuffleInfo = validateAndGetAppShuffleInfo(msg.appId);
     if (appShuffleInfo.attemptId != msg.appAttemptId) {
       // If this Block belongs to a former application attempt, it is considered late,
-      // as only the blocks from the current application attempt will be merged
-      // TODO: [SPARK-35548] Client should be updated to handle this error.
-      throw new IllegalArgumentException(
-        String.format("The attempt id %s in this FinalizeShuffleMerge message does not match "
-          + "with the current attempt id %s stored in shuffle service for application %s",
-          msg.appAttemptId, appShuffleInfo.attemptId, msg.appId));
+      // as only the blocks from the current application attempt will be merged.
+      // No need use callback to return to client now, because the finalizeShuffleMerge
+      // in DAGScheduler has no retry policy.
```

Comment:
Hmm...that's true, but then, `DAGScheduler` would fail due to timeout instead of the older attempt. It'd be better if we could return the exact failure cause.

Reply:
https://github.com/apache/spark/blob/bbf988bd73b00d18dd1d443f225b3915a2c4433f/common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockHandler.java#L218-L223

Took another look at this, although we don't explicitly use callback here, it seems like that `BlockPushNonFatalFailure` can't be caught at server side. So it will still return to `DAGScheduler` with the error.

Reply:
You are right @Ngone51 , i add the client catch to handle the  BlockPushNonFatalFailure in latest PR.

Reply:
@Ngone51 `ExternalBlockHandler` won't catch it since `BlockPushNonFatalFailure` is not an `IOException`. It would be caught in `TransportRequestHandler#processRpcRequest`.

https://github.com/apache/spark/blob/ccead315b30764d10d8356fb0d5fce459f6e4f93/common/network-common/src/main/java/org/apache/spark/network/server/TransportRequestHandler.java#L174-L177

If we want to do a similar handling for this, we would need to check the type of the exception there, and respond an `RPCResponse` if it's a `BlockPushNonFatalFailure`.
In addition, the `MergeFinalizerListener` cannot assume the response is always a `MergeStatuses` since it could also be a `BlockPushReturnCode`. So some proper handling there is also needed.

Reply:
Given the other comment from @mridulm about too old attempt only being seen by an already failed app attempt, I also don't feel that it's necessary to do the handling I mentioned earlier.

Reply:
> It would be caught in TransportRequestHandler#processRpcRequest.

I actually meant this too. @Victsm

Reply:
> Given the other comment from @mridulm about too old attempt only being seen by an already failed app attempt, I also don't feel that it's necessary to do the handling I mentioned earlier.

I was thinking about this too. So it might be ok to not handle it for simiplity. 

@zhuqi-lucas Sorry for the bother. Could you revert this part change?

Reply:
Thanks @Victsm @Ngone51 for clarify, i will revert the client catch in latest PR.